File: test/test_api_lines_collections.rb
require_relative "../lib/rgfa.rb"
require "test/unit"

TestAPI ||= Module.new
TestAPI::Lines ||= Module.new

class TestAPI::Lines::Collections < Test::Unit::TestCase

  def test_gfa1_collections
    gfa = RGFA.from_file("testdata/all_line_types.gfa1.gfa")
    # comments
    assert_equal(1, gfa.comments.size)
    assert(gfa.comments[0].content =~ /collections/)
    # containments
    assert_equal(2, gfa.containments.size)
    assert_equal(["2_to_6", "1_to_5"], gfa.containments.map(&:id))
    # dovetails
    assert_equal(4, gfa.dovetails.size)
    assert_equal(["1_to_2", "1_to_3", "11_to_12", "11_to_13"],
                 gfa.dovetails.map(&:id))
    # edges
    assert_equal(6, gfa.edges.size)
    assert_equal(["1_to_2", "1_to_3", "11_to_12",
                  "11_to_13", "2_to_6", "1_to_5"],
                 gfa.edges.map(&:id))
    # segments
    assert_equal([:"1", :"3", :"5", :"13", :"11", :"12", :"4", :"6", :"2"],
                 gfa.segments.map(&:name))
    # segment_names
    assert_equal([:"1", :"3", :"5", :"13", :"11", :"12", :"4", :"6", :"2"],
                 gfa.segment_names)
    # paths
    assert_equal([:"14", :"15"], gfa.paths.map(&:name))
    # path_names
    assert_equal([:"14", :"15"], gfa.path_names)
    # names
    assert_equal(gfa.segment_names + gfa.path_names,
                 gfa.names)
    # lines
    assert_equal(gfa.comments + gfa.headers + gfa.segments + gfa.edges +
                 gfa.paths, gfa.lines)
  end

  def test_gfa2_collections
    gfa = RGFA.from_file("testdata/all_line_types.gfa2.gfa")
    # comments
    assert_equal(3, gfa.comments.size)
    assert(gfa.comments[0].content =~ /collections/)
    # edges
    assert_equal([:"1_to_2", :"2_to_6", :"1_to_3",
                  :"11_to_12", :"11_to_13", :"1_to_5"],
                 gfa.edges.map(&:name))
    # edge_names
    assert_equal([:"1_to_2", :"2_to_6", :"1_to_3",
                  :"11_to_12", :"11_to_13", :"1_to_5"],
                 gfa.edge_names)
    # dovetails
    assert_equal([:"1_to_2", :"1_to_3", :"11_to_12", :"11_to_13"],
                 gfa.dovetails.map(&:name))
    # containments
    assert_equal([:"2_to_6", :"1_to_5"],
                 gfa.containments.map(&:name))
    # gaps
    assert_equal([:"1_to_11", :"2_to_12"], gfa.gaps.map(&:name))
    # gap_names
    assert_equal([:"1_to_11", :"2_to_12"], gfa.gap_names)
    # sets
    assert_equal([:"16", :"16sub"], gfa.sets.map(&:name))
    # set_names
    assert_equal([:"16", :"16sub"], gfa.set_names)
    # paths
    assert_equal([:"14", :"15"], gfa.paths.map(&:name))
    # path_names
    assert_equal([:"14", :"15"], gfa.path_names)
    # segments
    assert_equal([:"1", :"3", :"5", :"13", :"11", :"12", :"4", :"6", :"2"],
                 gfa.segments.map(&:name))
    # segment_names
    assert_equal([:"1", :"3", :"5", :"13", :"11", :"12", :"4", :"6", :"2"],
                 gfa.segment_names)
    # fragments
    assert_equal(["read1_in_2", "read2_in_2"], gfa.fragments.map(&:id))
    # external_names
    assert_equal([:"read1", :"read2"], gfa.external_names)
    # custom_record_keys
    assert_equal([:X, :Y], gfa.custom_record_keys)
    # custom_records
    assert_equal(3, gfa.custom_records.size)
    assert_equal([:X, :X, :Y], gfa.custom_records.map(&:record_type))
    # custom_records(:X)
    assert_equal([:X, :X], gfa.custom_records(:X).map(&:record_type))
    # names
    assert_equal(gfa.segment_names + gfa.edge_names + gfa.gap_names +
                 gfa.path_names + gfa.set_names, gfa.names)
    # lines
    assert_equal(gfa.comments + gfa.headers + gfa.segments + gfa.edges +
                 gfa.paths + gfa.sets + gfa.gaps + gfa.fragments +
                 gfa.custom_records, gfa.lines)
  end
end

File: tests/test_api_lines_collections.py
import gfapy
import unittest

class TestAPILinesCollections(unittest.TestCase):

  def test_gfa1_collections(self):
    gfa = gfapy.Gfa.from_file("tests/testdata/all_line_types.gfa1.gfa")
    # comments
    self.assertEqual(1, len(gfa.comments))
    self.assertRegex(gfa.comments[0].content, r'collections')
    # containments
    self.assertEqual(2, len(gfa.containments))
    self.assertEqual({"2_to_6", "1_to_5"}, {x.name for x in gfa.containments})
    # dovetails
    self.assertEqual(4, len(gfa.dovetails))
    self.assertEqual(set(["1_to_2", "1_to_3", "11_to_12", "11_to_13"]),
                 set([x.name for x in gfa.dovetails]))
    # edges
    self.assertEqual(6, len(gfa.edges))
    self.assertEqual(set(["1_to_2", "1_to_3", "11_to_12",
                  "11_to_13", "2_to_6", "1_to_5"]),
                 set([x.name for x in gfa.edges]))
    # segments
    self.assertSetEqual(set(["1", "3", "5", "13", "11", "12", "4", "6", "2"]),
                 set([x.name for x in gfa.segments]))
    # segment_names
    self.assertSetEqual(set(["1", "3", "5", "13", "11", "12", "4", "6", "2"]),
                 set(gfa.segment_names))
    # paths
    self.assertSetEqual(set(["14", "15"]), set([x.name for x in gfa.paths]))
    # path_names
    self.assertSetEqual(set(["14", "15"]), set(gfa.path_names))
    # names
    self.assertSetEqual(set(gfa.segment_names + gfa.path_names +
                        gfa.edge_names), set(gfa.names))
    # lines
    self.assertEqual(set([str(x) for x in gfa.comments + gfa.headers + gfa.segments + gfa.edges +
                 gfa.paths]), set([str(x) for x in gfa.lines]))

  def test_gfa2_collections(self):
    gfa = gfapy.Gfa.from_file("tests/testdata/all_line_types.gfa2.gfa")
    # comments
    self.assertEqual(3, len(gfa.comments))
    self.assertRegex(gfa.comments[0].content, r'collections')
    # edges
    self.assertSetEqual(set(["1_to_2", "2_to_6", "1_to_3",
                  "11_to_12", "11_to_13", "1_to_5"]),
                 set([x.name for x in gfa.edges]))
    # edge_names
    self.assertSetEqual(set(["1_to_2", "2_to_6", "1_to_3",
                  "11_to_12", "11_to_13", "1_to_5"]),
                 set(gfa.edge_names))
    # dovetails
    self.assertSetEqual(set(["1_to_2", "1_to_3", "11_to_12", "11_to_13"]),
                 set([x.name for x in gfa.dovetails]))
    # containments
    self.assertSetEqual(set(["2_to_6", "1_to_5"]),
                 set([x.name for x in gfa.containments]))
    # gaps
    self.assertSetEqual(set(["1_to_11", "2_to_12"]), set([x.name for x in gfa.gaps]))
    # gap_names
    self.assertSetEqual(set(["1_to_11", "2_to_12"]), set(gfa.gap_names))
    # sets
    self.assertSetEqual(set(["16", "16sub"]), set([x.name for x in gfa.sets]))
    # set_names
    self.assertSetEqual(set(["16", "16sub"]), set(gfa.set_names))
    # paths
    self.assertSetEqual(set(["14", "15"]), set([x.name for x in gfa.paths]))
    # path_names
    self.assertSetEqual(set(["14", "15"]), set(gfa.path_names))
    # segments
    self.assertSetEqual(set(["1", "3", "5", "13", "11", "12", "4", "6", "2"]),
               set([x.name for x in gfa.segments]))
    # segment_names
    self.assertSetEqual(set(["1", "3", "5", "13", "11", "12", "4", "6", "2"]),
               set(gfa.segment_names))
    # fragments
    self.assertSetEqual(set(["read1_in_2", "read2_in_2"]),
        set([x.get("id") for x in gfa.fragments]))
    # external_names
    self.assertSetEqual(set(["read1", "read2"]), set(gfa.external_names))
    # custom_record_keys
    self.assertSetEqual(set(["X", "Y"]), set(gfa.custom_record_keys))
    # custom_records
    self.assertEqual(3, len(gfa.custom_records))
    self.assertSetEqual(set(["X", "X", "Y"]), set([x.record_type for x in gfa.custom_records]))
    # custom_records("X")
    self.assertSetEqual(set(["X", "X"]), set([x.record_type for x in gfa.custom_records_of_type("X")]))
    # names
    self.assertSetEqual(set(gfa.segment_names + gfa.edge_names + gfa.gap_names +
                 gfa.path_names + gfa.set_names), set(gfa.names))
    # lines
    self.assertSetEqual(set([str(x) for x in gfa.comments + gfa.headers + gfa.segments + gfa.edges +
                 gfa.paths + gfa.sets + gfa.gaps + gfa.fragments +
                 gfa.custom_records]), set([str(x) for x in gfa.lines]))
